src/librustdoc/passes/strip_hidden.rs

```diff
@@ -89,6 +108,34 @@ impl Stripper<'_, '_> {
 impl DocFolder for Stripper<'_, '_> {
     fn fold_item(&mut self, i: Item) -> Option<Item> {
         let has_doc_hidden = i.is_doc_hidden();
+
+        if let clean::ImportItem(clean::Import { source, .. }) = &i.kind {
+            if let Some(source_did) = source.did {
+                let import_def_id = i.def_id().unwrap().expect_local();
+                let reexports = reexport_chain(self.tcx, import_def_id, source_did);
+
+                // Check if any reexport in the chain has a hidden source
+                let has_hidden_source =
+                    reexports.iter().filter_map(|reexport| reexport.id()).any(|reexport_did| {
+                        self.all_items
+                            .iter()
+                            .find(|item| item.def_id() == Some(reexport_did))
+                            .and_then(|item| {
+                                if let clean::ItemKind::ImportItem(import) = &item.kind {
+                                    import.source.did.map(|did| self.tcx.is_doc_hidden(did))
+                                } else {
+                                    None
+                                }
+                            })
+                            .unwrap_or(false)
+                    });
```

Comment:
So if I understand correctly: `self.all_items` is actually only the crate-level items (and not all items), then you iterate on these top-level items to find `use` items and check if they are `doc(hidden)`. Is there a reason you want to only check `ImportItem`? What happens if you check all items (including the original one)?

And also, you used `is_doc_hidden`, so you already are aware of it, so why keeping `all_items`?

If my hypothesis is correct, your code could become:

```suggestion
                let has_hidden_source = reexports.iter()
                    .filter_map(|reexport| reexport.id())
                    .any(|reexport_did| self.tcx.is_doc_hidden(reexport_did));
```

Reply:
`collect_all_items` recursively collects the items in the module, I did this because I found that reexport would only parse the `use` statements and the sources would not be obtained. For example, 
```
#[doc(hidden)]
pub fn hidden_self() {}

#[doc(hidden)]
pub mod hidden_m {
    pub fn hidden_super() {}
}

pub use crate::hidden_m::hidden_super as hidden_super_reexport;
pub use crate::hidden_self as hidden_self_reexport;

#[doc(inline)]
pub use crate::hidden_m::hidden_super as hidden_super_reexport_inline;
#[doc(inline)]
pub use crate::hidden_self as hidden_self_reexport_inline;
```
In this code,  `reexport_chain()` give me just the defid of `pub use crate::hidden_self as hidden_self_reexport;
`,  which is `reexports: [Single(DefId(0:7 ~ rustdoc_hidden_test[8c93]::{use#2}))]`.
But I should get the source of it, which is `#[doc(hidden)]pub fn hidden_self() {}`.
So, I do it to track the source of a reexport.

Reply:
Sorry, I forgot to recurse and didn't notice that `import.source` is the most starting point of the export. I changed the logic as you suggested, checking the reexport chain and also checking the source.  Now its a clean logic, thanks.

```
                let has_hidden_source = reexports
                    .iter()
                    .filter_map(|reexport| reexport.id())
                    .any(|reexport_did| self.tcx.is_doc_hidden(reexport_did))
                    || self.tcx.is_doc_hidden(source_did);
```